File: apps/agent/app/agent/tools/check_availability.py
"""Tool contract and implementation: check_availability."""

from datetime import datetime, date, time, timedelta, timezone
from typing import Any
import asyncpg
from pydantic import BaseModel, Field
from app.config import settings
from app.db.client import with_clinic
from app.llm.base import ToolSpec

TOOL_NAME = "check_availability"
TOOL_DESCRIPTION = "Query available appointment calendar slots for a date range, dentist, or procedure."

# Asia/Karachi is UTC+5
PK_TZ = timezone(timedelta(hours=5))


class Slot(BaseModel):
    """Calendar appointment slot representation."""

    start_at: str
    end_at: str
    dentist_id: str
    dentist_name: str | None = None


class CheckAvailabilityInput(BaseModel):
    """Input parameters for check_availability tool."""

    date: str = Field(..., description="Target date in YYYY-MM-DD format (or 'today', 'upcoming')")
    dentist_id: str | None = Field(None, description="Optional UUID of specific dentist")
    procedure: str | None = Field(None, description="Optional procedure name or code")


class CheckAvailabilityOutput(BaseModel):
    """Output results for check_availability tool."""

    date: str
    available_slots: list[Slot] = Field(default_factory=list)


TOOL_SPEC = ToolSpec(
    name=TOOL_NAME,
    description=TOOL_DESCRIPTION,
    parameters=CheckAvailabilityInput.model_json_schema(),
)


def _parse_target_date(raw_date: str) -> date:
    """Resolve raw date string to a valid date object."""
    clean = raw_date.strip().lower()
    now_pk = datetime.now(PK_TZ).date()
    if clean in ["today", "now"]:
        return now_pk
    elif clean in ["tomorrow", "upcoming", "next"]:
        return now_pk + timedelta(days=1)
    try:
        return datetime.strptime(clean, "%Y-%m-%d").date()
    except ValueError:
        return now_pk + timedelta(days=1)


async def execute(input_data: CheckAvailabilityInput, context: dict[str, Any]) -> CheckAvailabilityOutput:
    """Generate available appointment slots excluding booked appointments."""
    clinic_id = context.get("clinic_id") or settings.default_clinic_id
    target_date = _parse_target_date(input_data.date)
    date_str = target_date.isoformat()

    # Define day window in Pakistan time (09:00 to 19:00)
    day_start_dt = datetime.combine(target_date, time(9, 0), tzinfo=PK_TZ)
    day_end_dt = datetime.combine(target_date, time(19, 0), tzinfo=PK_TZ)

    async def _query(conn: asyncpg.Connection) -> CheckAvailabilityOutput:
        # 1. Fetch practitioners
        if input_data.dentist_id:
            dentists = await conn.fetch(
                "SELECT id, full_name FROM users WHERE clinic_id = $1::uuid AND id = $2::uuid",
                clinic_id,
                input_data.dentist_id,
            )
        else:
            dentists = await conn.fetch(
                "SELECT id, full_name FROM users WHERE clinic_id = $1::uuid AND role IN ('dentist', 'owner')",
                clinic_id,
            )

        if not dentists:
            # Fallback placeholder practitioner if clinic has no dentist rows
            dentists = [{"id": "00000000-0000-0000-0000-000000000001", "full_name": "Clinic Duty Dentist"}]

        # 2. Fetch active appointments on target date
        booked_appts = await conn.fetch(
            """
            SELECT id, dentist_id, start_at, end_at, status
            FROM appointments
            WHERE clinic_id = $1::uuid
              AND status NOT IN ('cancelled', 'no_show')
              AND start_at >= $2
              AND start_at < $3
            """,
            clinic_id,
            day_start_dt,
            day_end_dt,
        )

        available_slots: list[Slot] = []

        # 3. Generate 30-minute candidate slots for each dentist
        slot_duration = timedelta(minutes=30)
        for d in dentists:
            d_id = str(d["id"])
            d_name = d["full_name"]

            # Filter existing bookings for this dentist
            d_appts = [
                (a["start_at"], a["end_at"])
                for a in booked_appts
                if str(a["dentist_id"]) == d_id
            ]

            curr_slot_start = day_start_dt
            while curr_slot_start + slot_duration <= day_end_dt:
                curr_slot_end = curr_slot_start + slot_duration

                # Check if slot overlaps any existing appointment
                overlap = any(
                    (curr_slot_start < a_end and curr_slot_end > a_start)
                    for a_start, a_end in d_appts
                )

                if not overlap:
                    available_slots.append(
                        Slot(
                            start_at=curr_slot_start.isoformat(),
                            end_at=curr_slot_end.isoformat(),
                            dentist_id=d_id,
                            dentist_name=d_name,
                        )
                    )

                curr_slot_start = curr_slot_end

        return CheckAvailabilityOutput(date=date_str, available_slots=available_slots)

    return await with_clinic(clinic_id, _query)

File: apps/agent/src/agent/tools/check-availability.ts
import { z } from 'zod';
import { and, eq, notInArray, gte, lt, inArray } from 'drizzle-orm';
import { appointments, users } from '@dental-pms/db/schema';
import type { ToolSpec } from '../../llm/base';
import { runInClinic, getClinicIdFromContext } from '../../db/context';

export const TOOL_NAME = 'check_availability';
export const TOOL_DESCRIPTION =
  'Query available appointment calendar slots for a date range, dentist, or procedure.';

export const checkAvailabilityInputSchema = z.object({
  date: z.string().describe("Target date in YYYY-MM-DD format (or 'today', 'upcoming')"),
  dentist_id: z.string().uuid().optional().describe('Optional UUID of specific dentist'),
  procedure: z.string().optional().describe('Optional procedure name or code'),
});

export type CheckAvailabilityInput = z.infer<typeof checkAvailabilityInputSchema>;

export interface Slot {
  start: string;
  end: string;
  start_at: string;
  end_at: string;
  dentist_id: string;
  dentist_name?: string | null;
}

export interface CheckAvailabilityOutput {
  date: string;
  available_slots: Slot[];
}

export const TOOL_SPEC: ToolSpec = {
  name: TOOL_NAME,
  description: TOOL_DESCRIPTION,
  parameters: {
    type: 'object',
    properties: {
      date: {
        type: 'string',
        description: "Target date in YYYY-MM-DD format (or 'today', 'upcoming')",
      },
      dentist_id: {
        type: 'string',
        description: 'Optional UUID of specific dentist',
      },
      procedure: {
        type: 'string',
        description: 'Optional procedure name or code',
      },
    },
    required: ['date'],
  },
};

/**
 * Formats a Date to YYYY-MM-DD in Asia/Karachi (+05:00).
 */
function getKarachiDateString(date: Date = new Date()): string {
  const formatter = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'Asia/Karachi',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  });
  return formatter.format(date);
}

export async function execute(
  input: CheckAvailabilityInput,
  context?: Record<string, unknown>
): Promise<CheckAvailabilityOutput> {
  const clinicId = getClinicIdFromContext(context);

  return await runInClinic(clinicId, async (tx) => {
    // 1. Resolve dentists to consider
    let activeDentists: Array<{ id: string; fullName: string }> = [];
    if (input.dentist_id) {
      const [dentist] = await tx
        .select({ id: users.id, fullName: users.fullName })
        .from(users)
        .where(and(eq(users.id, input.dentist_id), eq(users.clinicId, clinicId)))
        .limit(1);
      if (dentist) {
        activeDentists = [dentist];
      }
    }

    if (activeDentists.length === 0) {
      activeDentists = await tx
        .select({ id: users.id, fullName: users.fullName })
        .from(users)
        .where(
          and(
            eq(users.clinicId, clinicId),
            inArray(users.role, ['dentist', 'owner'])
          )
        );
    }

    // If still no dentist configured, return empty
    if (activeDentists.length === 0) {
      return {
        date: input.date,
        available_slots: [],
      };
    }

    // 2. Resolve target dates
    const datesToScan: string[] = [];
    const karachiTodayStr = getKarachiDateString();

    if (input.date === 'upcoming') {
      // Next 7 days starting tomorrow
      const baseDate = new Date();
      for (let i = 1; i <= 7; i++) {
        const d = new Date(baseDate.getTime() + i * 24 * 60 * 60 * 1000);
        datesToScan.push(getKarachiDateString(d));
      }
    } else if (input.date === 'today') {
      datesToScan.push(karachiTodayStr);
    } else {
      // Specific YYYY-MM-DD format
      datesToScan.push(input.date);
    }

    const availableSlots: Slot[] = [];

    // 3. Scan dates until 10 slots are collected
    for (const targetDate of datesToScan) {
      if (availableSlots.length >= 10) break;

      // Start of clinic hours: 09:00 PKT (04:00 UTC)
      // End of clinic hours: 19:00 PKT (14:00 UTC)
      const dayStart = new Date(`${targetDate}T09:00:00+05:00`);
      const dayEnd = new Date(`${targetDate}T19:00:00+05:00`);

      // Query appointments on this day
      const existingAppointments = await tx
        .select({
          startAt: appointments.startAt,
          endAt: appointments.endAt,
          dentistId: appointments.dentistId,
        })
        .from(appointments)
        .where(
          and(
            notInArray(appointments.status, ['cancelled', 'no_show']),
            gte(appointments.startAt, dayStart),
            lt(appointments.startAt, dayEnd),
            input.dentist_id ? eq(appointments.dentistId, input.dentist_id) : undefined
          )
        );

      // Generate 30-minute intervals
      let cursor = new Date(dayStart.getTime());
      while (cursor < dayEnd && availableSlots.length < 10) {
        const slotStart = new Date(cursor.getTime());
        const slotEnd = new Date(cursor.getTime() + 30 * 60 * 1000);

        for (const dentist of activeDentists) {
          if (availableSlots.length >= 10) break;

          const isOverlapping = existingAppointments.some((appt) => {
            if (appt.dentistId !== dentist.id) return false;
            const apptStart = new Date(appt.startAt).getTime();
            const apptEnd = new Date(appt.endAt).getTime();
            return slotStart.getTime() < apptEnd && slotEnd.getTime() > apptStart;
          });

          if (!isOverlapping) {
            const startIso = slotStart.toISOString();
            const endIso = slotEnd.toISOString();
            availableSlots.push({
              start: startIso,
              end: endIso,
              start_at: startIso,
              end_at: endIso,
              dentist_id: dentist.id,
              dentist_name: dentist.fullName,
            });
          }
        }

        cursor = new Date(cursor.getTime() + 30 * 60 * 1000);
      }
    }

    return {
      date: input.date,
      available_slots: availableSlots.slice(0, 10),
    };
  });
}
